### PR TITLE
Pagination for 1.6.4 controls screen

### PR DESCRIPTION
**Pull request by thecatcore**
1.6.4 Controls screen doesn't support dynamic amount of keybindings, this pr provides a replacement for it to fix this issue.

Vanilla Screen:
Screenshot from 2024-06-27 21-17-05

New Screen (Ignore inverted color) :
Screenshot from 2024-06-29 19-53-46
Screenshot from 2024-06-29 19-53-56
Screenshot from 2024-06-29 19-54-03
Screenshot from 2024-06-29 19-54-11

### CODE REVIEW

**Review by BluCobalt — APPROVED** [review]
very cool

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/OptimizedListWidget.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+	private int homeButtonId;
+	private int endButtonId;
+	protected int lastMouseX;
+	protected int lastMouseY;
+	private float field_1253 = -2.0F;
+	private float scrollSpeed;
+	private float scrollAmount;
+	private int selectedEntry = -1;
+	private long time;
+	private boolean field_1258 = true;
```
Comment (BoogieMonster1O1) [inline legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/OptimizedListWidget.java:50]: What are these names

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/DynamicControlsOptionsScreen.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.widget.ButtonWidget;
+import net.minecraft.client.option.GameOptions;
+import net.minecraft.client.option.KeyBinding;
+import net.minecraft.client.render.Tessellator;
+import net.minecraft.client.resource.language.I18n;
+import net.minecraft.util.Formatting;
+
+public class DynamicControlsOptionsScreen extends Screen {
```
Comment (BoogieMonster1O1) [inline legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/DynamicControlsOptionsScreen.java:30]: Are we overwriting the existing screen?

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/mixin/client/keybinding/SettingsScreenMixin.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import net.legacyfabric.fabric.impl.client.keybinding.DynamicControlsOptionsScreen;
+
+@Mixin(SettingsScreen.class)
+public class SettingsScreenMixin extends Screen {
+	@Shadow
+	@Final
+	private GameOptions options;
+
+	@Inject(method = "buttonClicked", at = @At("HEAD"), cancellable = true)
+	private void hijackControlScreenButton(ButtonWidget buttonWidget, CallbackInfo ci) {
```
Comment (BoogieMonster1O1) [inline legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/mixin/client/keybinding/SettingsScreenMixin.java:41]: This is very concerning. Why can’t we inject into the existing screen? And how would other mods be able to inject into it if we hijack it?

**Review by BoogieMonster1O1 — CHANGES_REQUESTED** [review]
I don’t think completely bypassing the vanilla controls screen is the right approach.

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/OptimizedListWidget.java (hunk 1 above)
Comment (thecatcore) [reply legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/OptimizedListWidget.java:50]: This class is a modified version of vanilla ListWidget with fixes related to keyboard and mouse inputs detection. I haven't renamed missing mappings names yet.

**Comment by thecatcore** [comment]
> I don’t think completely bypassing the vanilla controls screen is the right approach. 

What do you think is the right approach in this case then?
The current implementation modifies the screen quite heavily, it would be difficult to keep mixins to the vanilla screen working with it.

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/mixin/client/keybinding/SettingsScreenMixin.java (hunk 3 above)
Comment (thecatcore) [reply legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/mixin/client/keybinding/SettingsScreenMixin.java:41]: I guess we can inject in vanilla control screen constructor instead?

**Comment by thecatcore** [comment]
Update with Pagination system instead.
Screenshot from 2024-07-01 20-01-11
Screenshot from 2024-07-01 20-01-17

File: legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/FabricControlsScreenComponents.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	private static final Identifier BUTTON_TEX = new Identifier("legacy-fabric", "textures/gui/creative_buttons.png");
+	public static final int AMOUNT_PER_PAGE = 14;
+
+	public static class ControlsButtonWidget extends ButtonWidget {
+		ControlsScreenExtensions extensions;
+		ControlsOptionsScreen gui;
+		Type type;
+
+		public ControlsButtonWidget(int x, int y, Type type, ControlsScreenExtensions extensions) {
+			super(1000 + type.ordinal(), x, y, 20, 20, "");
```
Comment (BoogieMonster1O1) [inline legacy-fabric-keybindings-api-v1/1.6.4/src/main/java/net/legacyfabric/fabric/impl/client/keybinding/FabricControlsScreenComponents.java:38]:
```suggestion
			super(1000 + type.ordinal(), x, y, 20, 20, type.text);
```

**Review by BoogieMonster1O1 — COMMENTED** [review]
Nice. very simple implementation

**Pull request merged**